File: reporter/src/batcher.rs
use crate::errors::ReporterError;
use crate::ingress::IngressClient;
use crate::models::TestcaseRequest;
use std::collections::VecDeque;
use std::sync::Arc;
use std::sync::atomic::Ordering;
use std::time::Duration;
use tokio::runtime::Runtime;
use tokio::sync::mpsc;
use tokio::time::{self, Instant};

enum BatcherMesssage {
    Testcase(TestcaseRequest),
    Shutdown,
}

pub struct Batcher {
    sender: mpsc::Sender<BatcherMesssage>,
    is_accepting: Arc<std::sync::atomic::AtomicBool>,
    worker_handle: Option<tokio::task::JoinHandle<()>>,
    errors: Arc<tokio::sync::Mutex<VecDeque<ReporterError>>>,
}

impl Batcher {
    pub fn new(runtime: Arc<Runtime>, ingress: Arc<IngressClient>) -> Self {
        let (sender, mut receiver) = mpsc::channel::<BatcherMesssage>(1000);
        let is_accepting = Arc::new(std::sync::atomic::AtomicBool::new(true));
        let errors = Arc::new(tokio::sync::Mutex::new(VecDeque::new()));

        let worker_handle = runtime.spawn({
            let errors = errors.clone();

            async move {
                let mut batch = Vec::<TestcaseRequest>::new();
                let mut last_send = Instant::now();
                let batch_timeout = Duration::from_secs(5);
                let max_batch_size = 100;

                loop {
                    tokio::select! {
                        Some(msg) = receiver.recv() => {
                            match msg {
                                BatcherMesssage::Testcase(testcase) => {
                                    batch.push(testcase);
                                    if batch.len() >= max_batch_size {
                                        if let Err(e) = ingress.create_testcases(std::mem::take(&mut batch)).await {
                                            let mut errors_guard = errors.lock().await;
                                            errors_guard.push_back(e);
                                        }
                                        last_send = Instant::now();
                                    }
                                }
                                BatcherMesssage::Shutdown => {
                                    if !batch.is_empty() {
                                        if let Err(e) = ingress.create_testcases(std::mem::take(&mut batch)).await {
                                            let mut errors_guard = errors.lock().await;
                                            errors_guard.push_back(e);
                                        }
                                    }
                                    break;
                                }
                            }
                        }
                        _ = time::sleep_until(last_send + batch_timeout) => {
                            if !batch.is_empty() {
                                if let Err(e) = ingress.create_testcases(std::mem::take(&mut batch)).await {
                                    let mut errors_guard = errors.lock().await;
                                    errors_guard.push_back(e);
                                }
                                last_send = Instant::now();
                            }
                        }
                    }
                }
            }
        });

        Self {
            sender,
            is_accepting,
            worker_handle: Some(worker_handle),
            errors,
        }
    }

    pub async fn add(&self, testcase: TestcaseRequest) -> Result<(), ReporterError> {
        if self.is_accepting.load(Ordering::SeqCst) {
            self.sender
                .send(BatcherMesssage::Testcase(testcase))
                .await
                .map_err(|e| {
                    ReporterError::Unknown(format!(
                        "error sending testcase to batcher queue: {}",
                        e
                    ))
                })?;
        }
        Ok(())
    }

    pub async fn shutdown(&mut self) -> Result<(), ReporterError> {
        self.is_accepting.store(false, Ordering::SeqCst);
        self.sender
            .send(BatcherMesssage::Shutdown)
            .await
            .map_err(|e| {
                ReporterError::Unknown(format!("error sending shutdown to batcher queue: {}", e))
            })?;
        if let Some(worker_handle) = self.worker_handle.take() {
            match worker_handle.await {
                Ok(_) => Ok(()),
                Err(e) => Err(ReporterError::Unknown(format!(
                    "error joining batcher worker thread: {}",
                    e
                ))),
            }
        } else {
            Ok(())
        }
    }

    pub async fn pop_error(&self) -> Option<ReporterError> {
        let mut errors_guard = self.errors.lock().await;
        errors_guard.pop_front()
    }
}

File: reporter/batcher.go
package reporter

import (
	"sync"
	"sync/atomic"
	"time"
)

const (
	batcherChannelCap = 100
	batchSize         = 100
	batchTimeout      = 5 * time.Second
)

type batcherMessage struct {
	testcase *TestcaseRequest
	shutdown bool
}

type Batcher struct {
	ingress     *IngressClient
	messages    chan batcherMessage
	isAccepting atomic.Bool
	errors      []error
	errorsMu    sync.Mutex
	done        chan struct{}
	wg          sync.WaitGroup
}

func NewBatcher(ingress *IngressClient) *Batcher {
	b := &Batcher{
		ingress:  ingress,
		messages: make(chan batcherMessage, batcherChannelCap),
		errors:   make([]error, 0),
		done:     make(chan struct{}),
	}
	b.isAccepting.Store(true)

	b.wg.Add(1)
	go b.worker()

	return b
}

func (b *Batcher) worker() {
	defer b.wg.Done()

	batch := make([]TestcaseRequest, 0, batchSize)
	ticker := time.NewTicker(batchTimeout)
	defer ticker.Stop()

	for {
		select {
		case msg := <-b.messages:
			if msg.shutdown {
				if len(batch) > 0 {
					if err := b.ingress.CreateTestcases(batch); err != nil {
						b.addError(err)
					}
				}
				return
			}

			if msg.testcase != nil {
				batch = append(batch, *msg.testcase)
				if len(batch) >= batchSize {
					if err := b.ingress.CreateTestcases(batch); err != nil {
						b.addError(err)
					}
					batch = make([]TestcaseRequest, 0, batchSize)
					ticker.Reset(batchTimeout)
				}
			}

		case <-ticker.C:
			if len(batch) > 0 {
				if err := b.ingress.CreateTestcases(batch); err != nil {
					b.addError(err)
				}
				batch = make([]TestcaseRequest, 0, batchSize)
			}
		}
	}
}

func (b *Batcher) addError(err error) {
	b.errorsMu.Lock()
	defer b.errorsMu.Unlock()
	b.errors = append(b.errors, err)
}

func (b *Batcher) Add(testcase TestcaseRequest) error {
	if !b.isAccepting.Load() {
		return nil
	}

	select {
	case b.messages <- batcherMessage{testcase: &testcase}:
		return nil
	default:
		return NewUnknownError("batcher queue is full")
	}
}

func (b *Batcher) Shutdown() error {
	b.isAccepting.Store(false)
	b.messages <- batcherMessage{shutdown: true}
	b.wg.Wait()
	close(b.done)
	return nil
}

func (b *Batcher) PopError() error {
	b.errorsMu.Lock()
	defer b.errorsMu.Unlock()

	if len(b.errors) == 0 {
		return nil
	}

	err := b.errors[0]
	b.errors = b.errors[1:]
	return err
}
